File: apps/api/app/models/workflow.py
"""Pydantic models for workflow engine operations."""
from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# Application Models
class ApplicationCreate(BaseModel):
    """Request model for creating application."""

    resume_id: UUID
    job_id: Optional[UUID] = None
    company_id: Optional[UUID] = None
    status: str = "wishlist"
    priority: str = "medium"
    probability: int = 50
    notes: Optional[str] = None
    metadata: Optional[dict[str, Any]] = None


class ApplicationUpdate(BaseModel):
    """Request model for updating application."""

    status: Optional[str] = None
    priority: Optional[str] = None
    probability: Optional[int] = None
    notes: Optional[str] = None
    metadata: Optional[dict[str, Any]] = None


class ApplicationResponse(BaseModel):
    """Response model for application."""

    id: UUID
    user_id: UUID
    resume_id: UUID
    job_id: Optional[UUID] = None
    company_id: Optional[UUID] = None
    status: str
    stage: Optional[str] = None
    priority: str
    probability: int
    notes: Optional[str] = None
    extra_metadata: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Task Models
class TaskCreate(BaseModel):
    """Request model for creating task."""

    application_id: UUID
    title: str
    description: Optional[str] = None
    due_date: Optional[datetime] = None
    priority: str = "medium"
    task_type: str = "general"


class TaskUpdate(BaseModel):
    """Request model for updating task."""

    title: Optional[str] = None
    description: Optional[str] = None
    due_date: Optional[datetime] = None
    priority: Optional[str] = None
    status: Optional[str] = None


class TaskResponse(BaseModel):
    """Response model for task."""

    id: UUID
    application_id: UUID
    user_id: UUID
    title: str
    description: Optional[str] = None
    due_date: Optional[datetime] = None
    priority: str
    status: str
    task_type: str
    extra_metadata: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Note Models
class NoteCreate(BaseModel):
    """Request model for creating note."""

    application_id: UUID
    content: str
    note_type: str = "general"


class NoteResponse(BaseModel):
    """Response model for note."""

    id: UUID
    application_id: UUID
    user_id: UUID
    content: str
    note_type: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Timeline Event Models
class TimelineEventCreate(BaseModel):
    """Request model for creating timeline event."""

    application_id: UUID
    event_type: str
    title: str
    description: Optional[str] = None
    event_date: Optional[datetime] = None
    extra_metadata: Optional[dict[str, Any]] = None


class TimelineEventResponse(BaseModel):
    """Response model for timeline event."""

    id: UUID
    application_id: UUID
    event_type: str
    title: str
    description: Optional[str] = None
    event_date: datetime
    extra_metadata: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Notification Models
class NotificationCreate(BaseModel):
    """Request model for creating notification."""

    notification_type: str
    title: str
    message: str
    priority: str = "medium"
    application_id: Optional[UUID] = None
    action_url: Optional[str] = None


class NotificationResponse(BaseModel):
    """Response model for notification."""

    id: UUID
    user_id: UUID
    application_id: Optional[UUID] = None
    notification_type: str
    title: str
    message: str
    priority: str
    is_read: bool
    action_url: Optional[str] = None
    extra_metadata: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Company Profile Models
class CompanyProfileCreate(BaseModel):
    """Request model for creating company profile."""

    name: str
    website: Optional[str] = None
    industry: Optional[str] = None
    size: Optional[str] = None
    location: Optional[str] = None
    description: Optional[str] = None
    notes: Optional[str] = None


class CompanyProfileUpdate(BaseModel):
    """Request model for updating company profile."""

    name: Optional[str] = None
    website: Optional[str] = None
    industry: Optional[str] = None
    size: Optional[str] = None
    location: Optional[str] = None
    description: Optional[str] = None
    notes: Optional[str] = None


class CompanyProfileResponse(BaseModel):
    """Response model for company profile."""

    id: UUID
    user_id: UUID
    name: str
    website: Optional[str] = None
    industry: Optional[str] = None
    size: Optional[str] = None
    location: Optional[str] = None
    description: Optional[str] = None
    notes: Optional[str] = None
    extra_metadata: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Workflow Engine Models
class WorkflowTriggerRequest(BaseModel):
    """Request model for triggering workflow."""

    event_type: str
    context: dict[str, Any]


class WorkflowTriggerResponse(BaseModel):
    """Response model for workflow trigger."""

    event_type: str
    actions_executed: int
    results: list[dict[str, Any]]


# Smart Automation Models
class UserActivityData(BaseModel):
    """Model for user activity data."""

    applications: list[dict[str, Any]] = []
    tasks: list[dict[str, Any]] = []
    job_matches: list[dict[str, Any]] = []
    skills: list[str] = []
    job_requirements: list[dict[str, Any]] = []
    last_activity_date: Optional[datetime] = None
    interviews: list[dict[str, Any]] = []
    career_health: dict[str, Any] = {}


class AutomationAnalysisResponse(BaseModel):
    """Response model for automation analysis."""

    total_recommendations: int
    recommendations: list[dict[str, Any]]
    generated_at: datetime


class MorningBriefResponse(BaseModel):
    """Response model for morning brief."""

    greeting: str
    today_priorities: list[dict[str, Any]]
    upcoming_interviews: list[dict[str, Any]]
    deadlines_today: list[dict[str, Any]]
    recommendations: list[dict[str, Any]]
    career_health: dict[str, Any]


# Dashboard Models
class DashboardMetrics(BaseModel):
    """Dashboard metrics."""

    total_applications: int
    applications_by_stage: dict[str, int]
    pending_tasks: int
    upcoming_interviews: int
    unread_notifications: int
    weekly_progress: dict[str, Any]
    response_rate: float
    interview_rate: float
    offer_rate: float


class DashboardResponse(BaseModel):
    """Response model for dashboard."""

    metrics: DashboardMetrics
    morning_brief: Optional[MorningBriefResponse] = None
    recent_activity: list[dict[str, Any]]
    top_recommendations: list[dict[str, Any]]


# Analytics Models
class AnalyticsMetrics(BaseModel):
    """Analytics metrics."""

    application_funnel: dict[str, int]
    success_rate: float
    response_rate: float
    interview_conversion: float
    offer_conversion: float
    technology_trends: dict[str, int]
    most_successful_resume: Optional[dict[str, Any]]
    most_successful_category: Optional[str]
    most_requested_skills: list[str]
    weak_areas: list[str]


class AnalyticsResponse(BaseModel):
    """Response model for analytics."""

    metrics: AnalyticsMetrics
    generated_at: datetime


# Kanban Models
class KanbanColumn(BaseModel):
    """Kanban column."""

    id: str
    title: str
    status: str
    applications: list[ApplicationResponse]


class KanbanBoardResponse(BaseModel):
    """Response model for kanban board."""

    columns: list[KanbanColumn]


# Career Case Models
class CareerCaseResponse(BaseModel):
    """Response model for career case."""

    application: ApplicationResponse
    resume: dict[str, Any]
    job: dict[str, Any]
    company: Optional[CompanyProfileResponse]
    timeline: list[TimelineEventResponse]
    tasks: list[TaskResponse]
    notes: list[NoteResponse]
    documents: list[dict[str, Any]]
    ai_insights: dict[str, Any]

File: apps/web/src/hooks/api/workflow.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import apiClient from '@/lib/api-client';

// Types
export interface Application {
  id: string;
  user_id: string;
  resume_id: string;
  job_id: string;
  company_id?: string;
  status: string;
  stage?: string;
  priority: string;
  probability: number;
  notes?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface Task {
  id: string;
  application_id: string;
  user_id: string;
  title: string;
  description?: string;
  due_date?: string;
  priority: string;
  status: string;
  task_type: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface TimelineEvent {
  id: string;
  application_id: string;
  event_type: string;
  title: string;
  description?: string;
  event_date: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface Notification {
  id: string;
  user_id: string;
  application_id?: string;
  notification_type: string;
  title: string;
  message: string;
  priority: string;
  is_read: boolean;
  action_url?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface CompanyProfile {
  id: string;
  user_id: string;
  name: string;
  website?: string;
  industry?: string;
  size?: string;
  location?: string;
  description?: string;
  notes?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface DashboardMetrics {
  total_applications: number;
  applications_by_stage: Record<string, number>;
  pending_tasks: number;
  upcoming_interviews: number;
  unread_notifications: number;
  weekly_progress: Record<string, any>;
  response_rate: number;
  interview_rate: number;
  offer_rate: number;
}

export interface MorningBrief {
  greeting: string;
  today_priorities: Array<{
    title: string;
    priority: string;
    estimated_time: string;
  }>;
  upcoming_interviews: Array<any>;
  deadlines_today: Array<any>;
  recommendations: Array<any>;
  career_health: Record<string, any>;
}

export interface DashboardData {
  metrics: DashboardMetrics;
  morning_brief?: MorningBrief;
  recent_activity: Array<{
    action: string;
    entity_type: string;
    created_at: string;
  }>;
  top_recommendations: Array<any>;
}

export interface KanbanColumn {
  id: string;
  title: string;
  status: string;
  applications: Application[];
}

export interface KanbanBoard {
  columns: KanbanColumn[];
}

// API Hooks
export function useApplications(status?: string) {
  return useQuery({
    queryKey: ['applications', status],
    queryFn: () => apiClient.get<Application[]>(`/api/v1/workflow/applications${status ? `?status=${status}` : ''}`),
  });
}

export function useApplication(applicationId: string) {
  return useQuery({
    queryKey: ['application', applicationId],
    queryFn: () => apiClient.get<Application>(`/api/v1/workflow/applications/${applicationId}`),
    enabled: !!applicationId,
  });
}

export function useCreateApplication() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: {
      resume_id: string;
      job_id: string;
      company_id?: string;
      status?: string;
      priority?: string;
      probability?: number;
      notes?: string;
      metadata?: Record<string, any>;
    }) => apiClient.post<Application>('/api/v1/workflow/applications', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['applications'] });
    },
  });
}

export function useUpdateApplication() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ applicationId, data }: {
      applicationId: string;
      data: {
        status?: string;
        priority?: string;
        probability?: number;
        notes?: string;
        metadata?: Record<string, any>;
      };
    }) => apiClient.patch<Application>(`/api/v1/workflow/applications/${applicationId}`, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['applications'] });
    },
  });
}

export function useTasks(applicationId?: string, status?: string) {
  return useQuery({
    queryKey: ['tasks', applicationId, status],
    queryFn: () => apiClient.get<Task[]>(`/api/v1/workflow/tasks${applicationId ? `?application_id=${applicationId}` : ''}${status ? `&status=${status}` : ''}`),
  });
}

export function useCreateTask() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: {
      application_id: string;
      title: string;
      description?: string;
      due_date?: string;
      priority?: string;
      task_type?: string;
    }) => apiClient.post<Task>('/api/v1/workflow/tasks', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['tasks'] });
    },
  });
}

export function useUpdateTask() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ taskId, data }: {
      taskId: string;
      data: {
        title?: string;
        description?: string;
        due_date?: string;
        priority?: string;
        status?: string;
      };
    }) => apiClient.patch<Task>(`/api/v1/workflow/tasks/${taskId}`, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['tasks'] });
    },
  });
}

export function useTimelineEvents(applicationId: string) {
  return useQuery({
    queryKey: ['timeline', applicationId],
    queryFn: () => apiClient.get<TimelineEvent[]>(`/api/v1/workflow/applications/${applicationId}/timeline`),
    enabled: !!applicationId,
  });
}

export function useCreateTimelineEvent() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: {
      application_id: string;
      event_type: string;
      title: string;
      description?: string;
      event_date?: string;
      metadata?: Record<string, any>;
    }) => apiClient.post<TimelineEvent>(`/api/v1/workflow/applications/${data.application_id}/timeline`, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['timeline'] });
    },
  });
}

export function useNotifications(unreadOnly: boolean = false) {
  return useQuery({
    queryKey: ['notifications', unreadOnly],
    queryFn: () => apiClient.get<Notification[]>(`/api/v1/workflow/notifications${unreadOnly ? '?unread_only=true' : ''}`),
  });
}

export function useMarkNotificationRead() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (notificationId: string) => apiClient.patch<Notification>(`/api/v1/workflow/notifications/${notificationId}/read`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications'] });
    },
  });
}

export function useCompanyProfiles() {
  return useQuery({
    queryKey: ['companies'],
    queryFn: () => apiClient.get<CompanyProfile[]>('/api/v1/workflow/companies'),
  });
}

export function useCreateCompanyProfile() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: {
      name: string;
      website?: string;
      industry?: string;
      size?: string;
      location?: string;
      description?: string;
    }) => apiClient.post<CompanyProfile>('/api/v1/workflow/companies', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['companies'] });
    },
  });
}

export function useDashboard() {
  return useQuery({
    queryKey: ['dashboard'],
    queryFn: () => apiClient.get<DashboardData>('/api/v1/workflow/dashboard'),
  });
}

export function useMorningBrief() {
  return useQuery({
    queryKey: ['morning-brief'],
    queryFn: () => apiClient.get<MorningBrief>('/api/v1/workflow/dashboard/morning-brief'),
  });
}

export function useKanbanBoard() {
  return useQuery({
    queryKey: ['kanban'],
    queryFn: () => apiClient.get<KanbanBoard>('/api/v1/workflow/kanban'),
  });
}

export function useTriggerWorkflow() {
  return useMutation({
    mutationFn: (data: {
      event_type: string;
      context: Record<string, any>;
    }) => apiClient.post('/api/v1/workflow/workflow/trigger', data),
  });
}

export function useAnalyzeAutomation() {
  return useMutation({
    mutationFn: (data: {
      applications: Array<any>;
      tasks: Array<any>;
      job_matches: Array<any>;
      skills: string[];
      job_requirements: Array<any>;
      last_activity_date?: string;
      interviews: Array<any>;
      career_health: Record<string, any>;
    }) => apiClient.post('/api/v1/workflow/automation/analyze', data),
  });
}
